extensions/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/util/PathFilter.java

```diff
@@ -0,0 +1,152 @@
+                    if (matcher.match(part, path)) {
+                        return true;
+                    }
+                }
+                return ObjectHelper.isEmpty(includePatterns);
+            };
+        }
+        ;
+    }
+
+    public Predicate<String> asStringPredicate() {
+        return stringPredicate;
+    }
+
+    public Predicate<DotName> asDotNamePredicate() {
+        if (ObjectHelper.isEmpty(excludePatterns) && ObjectHelper.isEmpty(includePatterns)) {
+            return dotName -> true;
+        } else {
+            return dotName -> stringPredicate.test(dotName.toString().replace('.', '/'));
+        }
+    }
+
+    public Predicate<Path> asPathPredicate() {
+        if (ObjectHelper.isEmpty(excludePatterns) && ObjectHelper.isEmpty(includePatterns)) {
+            return path -> true;
+        } else {
+            return path -> stringPredicate.test(sanitize(path.toString()));
+        }
+    }
+
+    static String sanitize(String path) {
+        path = path.trim();
+        return (!path.isEmpty() && path.charAt(0) == '/')
+                ? path.substring(1)
+                : path;
+    }
+
+    public static class Builder {
+        private List<String> includePatterns = new ArrayList<String>();
+        private List<String> excludePatterns = new ArrayList<String>();
+
+        public Builder patterns(boolean isInclude, Collection<String> patterns) {
+            if (isInclude) {
+                include(patterns);
+            } else {
+                exclude(patterns);
+            }
+            return this;
+        }
+
+        public Builder include(String pattern) {
+            includePatterns.add(sanitize(pattern));
+            return this;
+        }
+
+        public Builder include(Collection<String> patterns) {
+            patterns.stream().map(PathFilter::sanitize).forEach(includePatterns::add);
+            return this;
+        }
+
+        public Builder include(Optional<? extends Collection<String>> patterns) {
+            patterns.ifPresent(ps -> include(ps));
+            return this;
+        }
+
+        public Builder exclude(String pattern) {
+            excludePatterns.add(sanitize(pattern));
+            return this;
+        }
+
+        public Builder exclude(Collection<String> patterns) {
+            patterns.stream().map(PathFilter::sanitize).forEach(excludePatterns::add);
+            return this;
+        }
+
+        public Builder exclude(Optional<? extends Collection<String>> patterns) {
+            patterns.ifPresent(ps -> exclude(ps));
+            return this;
+        }
+
+        public Builder combine(Builder other) {
+            includePatterns.addAll(other.includePatterns);
+            excludePatterns.addAll(other.excludePatterns);
+            return this;
+        }
+
+        public PathFilter build() {
+            final List<String> incl = includePatterns;
+            includePatterns = null; // avoid leaking the collection trough reuse of the builder
```

Comment:
not a huge issue in our case but my understanding is that invoking `build()` multiple time on the same `builder` should produce the same result

Reply:
> invoking `build()` multiple time on the same `builder` should produce the same result

If the collection is not set to null here the following sequence would be possible:

```
PathFilter.Builder b = new PathFilter.Builder();
b.include("foo/bar");
PathFilter pf1 = b.build();
b.include("bar/baz");
PathFilter pf2 = b.build();
```

where pf1 and pf2 would have set the same list instance in their includePatterns fields. It find that really unwanted. That's a true reference leak that harms the immutability of PathFilter.

I hold re-using builders for rather unusual and so I find breaking the re-use better than copying the list in the PathFilter constructor as a way to ensure immutability.

Reply:
At least document it as I think no-one expect the builder to be re-set once you invoke `build()`

Reply:
Done at https://github.com/apache/camel-quarkus/pull/618/files#diff-7d0dfbc8ca22f0eb14e77322fcda0902R143

Reply:
Well another, in my opinion, cleaner option is to copy the collection in each object built making it immutable. A variation of this is used in most of the framework that generates builders, like Lombok and similars one.

My 2c.